x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractLogFileStructureFinder.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.logstructure;
+
+import org.elasticsearch.grok.Grok;
+import org.elasticsearch.xpack.ml.logstructure.TimestampFormatFinder.TimestampMatch;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractLogFileStructureFinder {
+
+    public static final String DEFAULT_TIMESTAMP_FIELD = "@timestamp";
+    public static final String MAPPING_TYPE_SETTING = "type";
+    public static final String MAPPING_FORMAT_SETTING = "format";
+    public static final String MAPPING_PROPERTIES_SETTING = "properties";
+
+    // NUMBER Grok pattern doesn't support scientific notation, so we extend it
+    private static final Grok NUMBER_GROK = new Grok(Grok.getBuiltinPatterns(), "^%{NUMBER}(?:[eE][+-]?[0-3]?[0-9]{1,2})?$");
+    private static final Grok IP_GROK = new Grok(Grok.getBuiltinPatterns(), "^%{IP}$");
+    private static final int KEYWORD_MAX_LEN = 256;
+    private static final int KEYWORD_MAX_SPACES = 5;
+
+    protected static Map<String, String> guessScalarMapping(List<String> explanation, String fieldName, Collection<String> fieldValues) {
```

Comment:
It would be nicer in some ways, but it would cause quite a lot of side effects.  At the moment `explanation` is built up by 3 pieces of code:

1. The character set finder
2. The high level structure determination (i.e. the `*LogFileStructure::canCreateFromSample()` methods)
3. The detailed structure determination

The unit tests test static methods that update explanations without having to create concrete objects.  And the unit tests also print the explanation or partial explanation for whatever they've tested, which means that for unit tests we want the explanation to be owned by the test code.

I think one solution would be to introduce an `Explanation` class that encapsulates the list of strings.  Then that could be a member of `AbstractLogFileStructureFinder`, which would avoid passing it around the member functions of that hierarchy so much.  But it would be documented that other classes might be adding to the same explanation.

Reply:
It's also tricky because the `GrokPatternCreator` needs to find mappings, relying on the fact that `guessScalarMapping()` is `static`.